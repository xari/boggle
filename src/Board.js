import React, { useEffect } from "react";
import { classNames } from "./utils";
import { reduceLetters, createRandomBoard } from "./utils";

export function Custom({ setBoard, dimensions, enableRandom }) {
  useEffect(() => {
    setBoard(null); // Clears the results whenever the board is resized

    return () => setBoard(createRandomBoard(dimensions)); // Generates a new random board on unmount
  }, [setBoard, dimensions]);

  const handleSubmit = (e) => {
    e.preventDefault(); // Avoid page refresh

    // Create an array of lower case letters from the input values
    const letters = [...e.target.elements.letter].map((inputNode) =>
      inputNode.value.toLowerCase()
    );

    const newBoard = reduceLetters(dimensions)(letters);

    setBoard(newBoard);
  };

  return (
    <form onSubmit={handleSubmit}>
      <div className={`grid grid-cols-${dimensions} gap-1`}>
        {[...Array(dimensions * dimensions).keys()].map((i) => (
          <div
            key={i}
            className={classNames(
              dimensions === 4
                ? "h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl"
                : dimensions === 5
                ? "h-20 sm:h-24 w-20 sm:w-24 sm:text-2xl md:text-3xl"
                : "h-16 sm:h-20 w-16 sm:w-20 sm:text-xl md:text-2xl",
              "flex p-3 border-2 content-center items-center text-center rounded"
            )}
          >
            <input
              type="text"
              name={`letter`}
              className="w-full font-light text-center focus:outline-none"
              maxLength="1"
              pattern="[A-Za-z]"
              required
              aria-label="Boggle letter"
            />
          </div>
        ))}
      </div>
      <input
        type="submit"
        className="my-3 inline-flex items-center px-6 py-3 border border-transparent text-base font-medium rounded-md shadow-sm text-white bg-emerald-600 hover:bg-emerald-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-emerald-500"
        value="Find all words"
        aria-label="Solve the board"
      />
    </form>
  );
}

export function Random({ board, setBoard, dimensions }) {
  useEffect(() => {
    setBoard(createRandomBoard(dimensions)); // Generate a new random board if the dimensions change

    return () => setBoard(null); // Clear the board on unmount
  }, [setBoard, dimensions]);

  return (
    <div className={`grid grid-cols-${dimensions} gap-1`}>
      {board &&
        board
          .flatMap((letter) => letter)
          .map((letter, i) => (
            <div
              key={i}
              className={classNames(
                dimensions === 4
                  ? "h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl"
                  : dimensions === 5
                  ? "h-20 sm:h-24 w-20 sm:w-24 sm:text-2xl md:text-3xl"
                  : "h-16 sm:h-20 w-16 sm:w-20 sm:text-xl md:text-2xl",
                "flex p-3 border-2 content-center items-center text-center rounded bg-neutral-50"
              )}
            >
              <span className="w-full font-light">{letter}</span>
            </div>
          ))}
    </div>
  );
}

export default function Board({ enableRandom, ...props }) {
  return (
    <div className="my-3 mx-auto flex">
      {enableRandom === true ? <Random {...props} /> : <Custom {...props} />}
    </div>
  );
}
