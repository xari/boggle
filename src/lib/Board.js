import React from "react";
import { classNames } from "./utils";

export default function Board({
  board,
  enabledRandom,
  setBoard,
  submitted,
  setSubmitted,
}) {
  const dimensions = Math.sqrt(board.length);
  const handleSubmit = (e) => {
    e.preventDefault(); // Avoid page refresh

    setSubmitted();
  };

  return (
    <div className="my-3 mx-auto flex">
      <form onSubmit={handleSubmit}>
        <fieldset disabled={enabledRandom ? "disabled" : ""}>
          <div className={`grid grid-cols-${dimensions} gap-1`}>
            {board.map((value, i, arr) => (
              <div
                key={i}
                className={classNames(
                  dimensions === 4
                    ? "h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl"
                    : dimensions === 5
                    ? "h-20 sm:h-24 w-20 sm:w-24 sm:text-2xl md:text-3xl"
                    : "h-16 sm:h-20 w-16 sm:w-20 sm:text-xl md:text-2xl",
                  "flex border-2 content-center items-center text-center rounded"
                )}
              >
                <input
                  type="text"
                  name="letter"
                  className={classNames(
                    enabledRandom && "bg-neutral-50",
                    "h-full w-full font-light text-center focus:outline-none"
                  )}
                  maxLength="1"
                  pattern="[A-Za-z]"
                  required
                  aria-label="Boggle letter"
                  value={value}
                  onChange={(e) => {
                    submitted === true && setSubmitted(true); // true means reset

                    setBoard([
                      ...arr.slice(0, i),
                      e.target.value.toLowerCase(),
                      ...arr.slice(i + 1),
                    ]);
                  }}
                />
              </div>
            ))}
          </div>
        </fieldset>
        <input
          type="submit"
          className="my-3 inline-flex items-center px-6 py-3 border border-transparent text-base font-medium rounded-md shadow-sm text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
          value="Find all words"
          aria-label="Solve the board"
        />
      </form>
    </div>
  );
}
