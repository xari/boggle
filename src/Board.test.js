import { shallow } from "enzyme";
import { Random } from "./Board";

describe("the Boggle board", () => {
  const board = [
    ["v", "z", "g", "x"],
    ["j", "z", "g", "b"],
    ["t", "b", "j", "r"],
    ["u", "a", "c", "p"],
  ];
  const gameWrapper = shallow(<Random board={board} />);
  const grid = gameWrapper.find(".grid");

  test("should be the correct length", () => {
    expect(grid.children()).toHaveLength(16);
  });

  test("should contain all of the letters", () => {
    const boardArr = board.flatMap((x) => x);

    grid
      .find("span")
      .forEach((node, i) => expect(node.text()).toEqual(boardArr[i]));
  });
});
