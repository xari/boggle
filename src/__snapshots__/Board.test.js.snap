// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Custom board renders without any unexpected changes 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="grid grid-cols-4 gap-1"
  >
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
    <div
      className="h-24 sm:h-28 w-24 sm:w-28 text-3xl sm:text-4xl flex p-3 border-2 content-center items-center text-center rounded"
    >
      <input
        aria-label="Boggle letter"
        className="w-full font-light text-center focus:outline-none"
        maxLength="1"
        name="letter"
        pattern="[A-Za-z]"
        required={true}
        type="text"
      />
    </div>
  </div>
  <input
    aria-label="Solve the board"
    className="my-3 inline-flex items-center px-6 py-3 border border-transparent text-base font-medium rounded-md shadow-sm text-white bg-emerald-600 hover:bg-emerald-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-emerald-500"
    type="submit"
    value="Find all words"
  />
</form>
`;
